### DISCUSSION TITLE
Ray Leakage in Indoor Scenarios with Sionna RT

### DISCUSSION THREAD

**Discussion by devrimfiratyilmaz** [NVlabs/sionna#564] [Show and tell]
While working with Sionna RT (v0.17.0) on my indoor scenarios, I have encountered an issue related to ray leakage at specific structural junctions. After converting a 2D floor plan to a 3D PLY format and defining the entire building with the 'mat-itu_marble' material, it was noted that non-white squares were present outside the building on the coverage maps.
![Sionna_Problem_1](https://github.com/user-attachments/assets/b770d9db-9756-40b3-b297-442697e6ae3f)

I suspected that these anomalies were caused by rays leaking outside, so I placed receivers outside the building to examine the paths in detail. My observations indicated that rays were consistently escaping the building at the junctions where the side walls meet the ceiling/floor.
![path_max_depth_15](https://github.com/user-attachments/assets/2604464f-6511-4712-a64f-d231171c2d0a)

Despite testing with different conversion methods from 2D to 3D and with the scene_01 of the published work WiSegRT, the issue remained unresolved. This indicated that the rays were still escaping from the building at the same points where the base and the side walls converge.
![Sionna_Problem_2](https://github.com/user-attachments/assets/5ebaa9ab-6702-49ee-8e55-da2851de61ba)

![Example scene _wiseg](https://github.com/user-attachments/assets/640f37d0-2b66-4441-a850-153b1f486452)

It seems that there may be an issue with how Sionna handles the intersections between side walls and ceilings. This is causing incorrect indoor/outdoor detection at these points, which is allowing rays to escape.

I would be grateful if you could suggest strategies or modifications to prevent rays from escaping at these junctures, thereby improving the accuracy of simulations.

**Comment by merlinND** [comment]
Hello @devrimfiratyilmaz,

I believe that OptiX, the library used for ray tracing in Sionna, can guarantee that rays won't leak this way only if the different parts are actually connected (i.e. part of the same mesh). Could you please confirm this is the case?
If not, you might have to very slightly overlap the edges of walls & ceilings to ensure there's no gap due to e.g. limited numerical precision.

**Reply by devrimfiratyilmaz** [reply]
Hello @merlinND,

Thank you so much for your feedback!

Each scene we have used in Sionna contains only a single PLY file, or in other words, a single mesh file. Based on your suggestion, we have made some progress and would like to give you an update:

- We noticed that a secondary surface was unintentionally overlapping with the rectangular shape we intended to define for both the floor and ceiling. After resolving this overlap (screenshots below and PLY files are attached [PLY_files.zip](https://github.com/user-attachments/files/16919322/PLY_files.zip)), we observed that the leaks originating from the floor and ceiling seem to have disappeared, according to our initial observations.
![Duplicate_surfaces](https://github.com/user-attachments/assets/815f0c93-6d17-41bb-8a78-fab4f3af2f4c)


- However, we are still seeing a few leaks, though their number has significantly decreased, and we haven't yet identified the exact cause.

- More importantly, when we continued our work with version 0.18.0 (our previous results were obtained with version 0.17.0), we observed a dramatic increase in ray leakage. To better visualize this, we increased the max_depth parameter of the coverage map from the default value to 30. Running the same notebook in an environment with Sionna v0.18.0 showed a striking increase in ray leakage compared to v0.17.0. Below, you can see the differences for three randomly selected scenes.
![Sionna_version_problem](https://github.com/user-attachments/assets/e6ee74bb-5c8b-4745-a939-f620e26b0599)

I would really appreciate your insights into the dramatic difference between v0.17.0 and v0.18.0 regarding the leakage issue, and I would be grateful for any further assistance you can provide.

**Reply by merlinND** [reply]
Hello @devrimfiratyilmaz,

Thank you for sharing the PLY files. I imported them in Blender and indeed, the floor & ceiling are *not* connected to the walls. As per my first message, there is no guarantee that rays cannot "slip through" unless you either connect them to the walls or slightly overlap the floor & ceiling with the walls.

In this video, you can see that the triangles forming the walls (purple) are connected to each other, but not to the ceiling (orange) or floor (green).
[Screencast from 2024-09-10 10-03-32.webm](https://github.com/user-attachments/assets/7ced0e41-04f6-4329-a728-b254378ce966)

**Reply by merlinND** [reply]
Regarding the comparison between v0.17 and v0.18, I agree there seems to be significantly more leakage, but also the coverage maps themselves look fairly different.
Are you sure that the exact same simulation & parameters are being run in the left and right columns?

If so, could you please share the full scenes (XML and PLY files) so that we can take a look.

**Reply by devrimfiratyilmaz** [reply]
Hello @merlinND ,

Thank you for your feedback.

When we look at a simplified scene that consists only of a single prism and use the exact same simulation and parameters (running the same notebook), we consistently observe significantly higher leakage in the coverage map with v0.18 compared to v0.17.
![comparision_v0 17_and_v0 18](https://github.com/user-attachments/assets/c1adc19a-ff11-442f-92ce-997e29c06271)

In this case, the content of the PLY file is as follows:
"
ply
format ascii 1.0
element vertex 16
property float x
property float y
property float z
element face 20
property list uchar int vertex_indices
end_header
0.11 0.1 0
0.1 0.7 0
4.61 0.7 0
4.61 0.1 0
0.11 0.1 3.0
0.1 0.7 3.0
4.61 0.7 3.0
4.61 0.1 3.0
-10.3 -10.3 0
15.010000000000002 -10.3 0
15.010000000000002 11.1 0
-10.3 11.1 0
-10.3 -10.3 3.0
15.010000000000002 -10.3 3.0
15.010000000000002 11.1 3.0
-10.3 11.1 3.0
3 0 1 5
3 0 5 4
3 1 2 6
3 1 6 5
3 2 3 7
3 2 7 6
3 3 0 4
3 3 4 7
3 8 9 10
3 8 10 11
3 12 13 14
3 12 14 15
3 8 9 13
3 8 13 12
3 9 10 14
3 9 14 13
3 10 11 15
3 10 15 14
3 11 8 12
3 11 12 15
"
The XML and PLY files are also attached ([example_scene.zip](https://github.com/user-attachments/files/16946159/example_scene.zip))

We didn't quite understand how to make the "connect" you mentioned. How should we modify the content of the PLY file in this simplified example? I would appreciate it if you could explain how to make the "connect" between the floor and ceiling in this or another PLY file example.

**Reply by devrimfiratyilmaz** [reply]
We have also attempted to slightly "overlap" the floor and ceiling with the walls by extending the walls by 0.5 units both upwards and downwards.

The content of the modified PLY file ([PLY.zip](https://github.com/user-attachments/files/16957526/PLY.zip)) is as follows:

ply
format ascii 1.0
element vertex 16
property float x
property float y
property float z
element face 20
property list uchar int vertex_indices
end_header
0.11 0.1 -0.5
0.1 0.7 -0.5
4.61 0.7 -0.5
4.61 0.1 -0.5
0.11 0.1 3.5
0.1 0.7 3.5
4.61 0.7 3.5
4.61 0.1 3.5
-10.3 -10.3 0
15.010000000000002 -10.3 0
15.010000000000002 11.1 0
-10.3 11.1 0
-10.3 -10.3 3.0
15.010000000000002 -10.3 3.0
15.010000000000002 11.1 3.0
-10.3 11.1 3.0
3 0 1 5
3 0 5 4
3 1 2 6
3 1 6 5
3 2 3 7
3 2 7 6
3 3 0 4
3 3 4 7
3 8 9 10
3 8 10 11
3 12 13 14
3 12 14 15
3 8 9 13
3 8 13 12
3 9 10 14
3 9 14 13
3 10 11 15
3 10 15 14
3 11 8 12
3 11 12 15
 
Unfortunately, this modification did not result in improved coverage maps.

![new_coverage_maps](https://github.com/user-attachments/assets/57be7b40-d348-43dd-8369-431475be2667)

I would appreciate it if you could review this and confirm whether our interpretation of "overlap" aligns with your expectations.

**Reply by faycalaa** [reply]
Hi @devrimfiratyilmaz,

We have reproduced the issue. It will be fixed in the next release.

**Reply by merlinND** [reply]
Thanks @devrimfiratyilmaz for providing the reproducers that helped figure out the source of the issue.

The "overlap" you added was indeed what I had in mind, but it didn't fix the problem because it's due to a recent code change instead of the specific geometry of your scene.

**Reply by devrimfiratyilmaz** [reply]
Thank you for your feedback.

At which version was the mentioned code change implemented? It cannot be the one introduced in v0.18.0, because you can also see the leaks in v0.17.0. We wanted to point this out again because a rollback of the last relevant changes from v0.17.0 to v0.18.0 would apparently not completely fix the problem.

**Reply by JackYifan** [reply]
@faycalaa May I ask when the new version of Sionna will be released?

**Reply by devrimfiratyilmaz** [reply]
Hello @faycalaa,

I have just completed testing the scenarios we discussed, now using the newly released version Sionna v0.19. Contrary to the announcement [here](https://github.com/NVlabs/sionna/discussions/605), while this version addresses the significant ray leakage issue present in v0.18, it does not fully resolve the leakage problem. In fact, the level of leakage observed is roughly similar to what was experienced in v0.17.

As we had expressed earlier, our concerns about rolling back the last relevant changes from v0.17.0 to v0.18.0 not completely addressing the issue seem valid. The reduced, yet persistent, leakage raises questions. As @merlinND mentioned in his response, if the errors are not related to the specific geometry of the scene, what could be causing them?

I would appreciate further insights into whether this might be related to other recent code changes or if additional fixes are planned for future versions. I look forward to your feedback.

**Reply by merlinND** [reply]
Hello @devrimfiratyilmaz,

Spawning reflected / refracted rays in a way that robustly avoids leakage with all types of geometry (including different scene scales) is far from a trivial problem due to the finite precision of floating point computations. See for example:
- https://www.pbr-book.org/4ed/Shapes/Managing_Rounding_Error#RobustSpawnedRayOrigins
- https://developer.nvidia.com/blog/solving-self-intersection-artifacts-in-directx-raytracing/

The fix mentioned in the v0.19 release notes meant fixing the regression that was introduced in v0.17.
We may introduce even more robust ray spawning algorithms in future versions.

**Reply by devrimfiratyilmaz** [reply]
Hello @merlinND,

Thank you for the clarification! We look forward to new developments in future versions!
